service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AgentDirectoryMapper.java

```diff
@@ -31,153 +26,57 @@
 public class AgentDirectoryMapper {
     private static final Mustache AGENT_DIRECTORY_STRUCTURE_TEMPLATE = TemplateUtils.loadTemplate("ehr_agent_structure_template.mustache");
 
-    private final PractitionerAgentPersonMapper practitionerAgentPersonMapper;
-    private final OrganizationToAgentMapper organizationToAgentMapper;
+    private final MessageContext messageContext;
+    private final AgentPersonMapper agentPersonMapper;
 
     public String mapEHRFolderToAgentDirectory(Bundle bundle, String nhsNumber) {
-        Set<String> mappedOrganizationsAndPractitioner = new HashSet<>();
-        var builder = AgentDirectoryParameter.builder()
-            .patientManagingOrganization(extractPatientManagingOrganization(bundle, nhsNumber))
-            .observationOrganizations(prepareObservationAgentsList(bundle, mappedOrganizationsAndPractitioner))
-            .referralRequestsOrganizations(prepareReferralRequestAgentsList(bundle, mappedOrganizationsAndPractitioner))
-            .agentPersons(preparePractitionerRoleAgentsList(bundle, mappedOrganizationsAndPractitioner));
+        var builder = AgentDirectoryParameter.builder();
 
-        return TemplateUtils.fillTemplate(AGENT_DIRECTORY_STRUCTURE_TEMPLATE, builder.build());
-    }
+        Set<Map.Entry<AgentDirectory.AgentKey, String>> agentDirectories = messageContext.getAgentDirectory().getEntries();
 
-    private String extractPatientManagingOrganization(Bundle bundle, String nhsNumber) {
-        return AgentDirectoryExtractor.extractPatientByNhsNumber(bundle, nhsNumber)
-            .filter(Patient::hasManagingOrganization)
-            .map(patient -> ResourceExtractor.extractResourceByReference(bundle,
-                patient.getManagingOrganization().getReferenceElement()))
-            .map(Optional::get)
-            .map(Organization.class::cast)
-            .map(organizationToAgentMapper::mapOrganizationToAgent)
+        Optional<Organization> patientManagingOrganization = extractPatientManagingOrganization(bundle, nhsNumber);
+
+        patientManagingOrganization
+            .map(organization -> OrganizationToAgentMapper.mapOrganizationToAgent(organization, getAgentId(organization)))
+            .map(builder::patientManagingOrganization)
             .orElseThrow(() -> new EhrMapperException("The ASR bundle does not contain a Patient resource with the correct identifier "
                 + "and a managingOrganization"));
```

Comment:
I found this a bit tricky to follow, I think because some of the ops work but aren't really thematic.

If I understand it right, the exception should be thrown if `patientManagingOrganization` comes back empty, rather than from the interim ops, so my first suggestion is to put the `.orElseThrow()` there. Then `patientManagingOrganization` is just an `Organization` rather than an `Optional` of one.

Then the builder can be invoked in a chain for the two properties, `.patientManagingOrganization()` and `.agentPersons()`.

```suggestion
        Organization patientManagingOrganization = extractPatientManagingOrganization(bundle, nhsNumber)
            .orElseThrow(() -> new EhrMapperException("The ASR bundle does not contain a Patient resource with the correct identifier "
                + "and a managingOrganization"));

        String agentId = getAgentId(patientManagingOrganization);
        builder.patientManagingOrganization(mapOrganizationToAgent(patientManagingOrganization, agentId))
            .agentPersons(agentDirectories.stream()
                .filter(agentDirectory -> !isPatientManagingOrganization(patientManagingOrganization, agentDirectory))
                .map(this::mapAgentDirectory)
                .collect(Collectors.toList())
            );
```

What do you think?

Reply:
you're right, I changes it a little different, please take a look.